### PR TITLE
Fix S6588 FN: Rule should cover case with epoch ticks

### LINKED ISSUE #7547: Fix S6588 FN: Rule should cover case with epoch ticks
### Description

The rule should raise if the constructor is called with the ticks that are equal to the epoch ticks (621355968000000000).

```cs
const long EpochTicks = 621355968000000000;

_ = new DateTime(621355968000000000); // Noncompliant
_ = new DateTime(EpochTicks); // Noncompliant
```

While doing the validation for rule S6562 I found examples like: _ = new DateTime(EPOCH);

### PR DESCRIPTION
**Pull request by cristian-ambrosini-sonarsource**
Fixes #7547

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.Common/Rules/UseUnixEpochBase.cs
Comment (zsolt-kolbay-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Rules/UseUnixEpochBase.cs:73]: I don't think this extra check is necessary. The previous check is more than enough. If someone gives the Epoch ticks in hexa or binary format, then I think we can live with that 😎

File: analyzers/src/SonarAnalyzer.Common/Rules/UseUnixEpochBase.cs
Comment (zsolt-kolbay-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Rules/UseUnixEpochBase.cs:64]: The two if statements can be merged if you move `CheckAndGetTypeName` to the end of the conditions.

File: analyzers/src/SonarAnalyzer.Common/Rules/UseUnixEpochBase.cs
Comment (zsolt-kolbay-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Rules/UseUnixEpochBase.cs:72]: This can also be done inside the `IsEpochCtor` method:

`|| IsParameterExistingAndLiteralEqualTo("ticks", EpochYear, lookup)`

**Review by zsolt-kolbay-sonarsource — APPROVED** [review]
LGTM
Added some polishing comments

File: analyzers/src/SonarAnalyzer.Common/Rules/UseUnixEpochBase.cs
Comment (cristian-ambrosini-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Rules/UseUnixEpochBase.cs:73]: The previous check is just for the hardcoded number:
```cs
_ = new DateTime(621355968000000000); // Noncompliant
```
while this check is actually looking for constants through the tracker:
```cs
private const long EpochTicks = 621355968000000000;
...
_ = new DateTime(EpochTicks);
```

File: analyzers/src/SonarAnalyzer.Common/Rules/UseUnixEpochBase.cs
Comment (csaba-sagi-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Rules/UseUnixEpochBase.cs:112]: NITPICK: You could return the string directly and  in the if check if  is {} typeName
```suggestion
    private static string CheckAndGetTypeName(SyntaxNode node, SemanticModel model)
```

File: analyzers/src/SonarAnalyzer.Common/Rules/UseUnixEpochBase.cs
Comment (csaba-sagi-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Rules/UseUnixEpochBase.cs:76]:
```suggestion
                        c.ReportIssue(Diagnostic.Create(Rule, c.Node.GetLocation(), typeName));
```

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/UseUnixEpoch.vb
Comment (csaba-sagi-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/UseUnixEpoch.vb:11]: I think you can add test cases with hex numbers and also maybe one with decimal numbers that contain underscores.
Both should work. It would be there just to document what we cover.

Comment (csaba-sagi-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/UseUnixEpoch.vb:11]: Same applies for both languages.

File: analyzers/src/SonarAnalyzer.Common/Rules/UseUnixEpochBase.cs
Comment (cristian-ambrosini-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Rules/UseUnixEpochBase.cs:64]: Unfortunately, the conditions are different: the first one is checking whether you have 3 literals (2 with number `1`, 1 with number `1970`) and then proceeds to check the type and constructor.
The second one is only searching for the constructor with a single argument of `621355968000000000` or pointing at a const with that value, and then proceeds to check semantically the type (not the constructor, since we have only 1 ctor with 1 argument for DateTime/DateTimeOffset).
Although it's feasible to merge everything in a single if statement, all the variants I tried are really cumbersome to read and understand. Happy to discuss it further

File: analyzers/src/SonarAnalyzer.Common/Rules/UseUnixEpochBase.cs
Comment (csaba-sagi-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Rules/UseUnixEpochBase.cs:36]: Why did you change this to hexa? In my opinion it is more understandable in decimal. We use that value more often than hexa numbers.

File: analyzers/src/SonarAnalyzer.Common/Rules/UseUnixEpochBase.cs
Comment (csaba-sagi-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Rules/UseUnixEpochBase.cs:37]: Why this comment at the year? That is the number of ticks. If we want to add that as a comment it should go with the ticks. However, I think it is not needed.

File: analyzers/src/SonarAnalyzer.Common/Rules/UseUnixEpochBase.cs
Comment (csaba-sagi-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Rules/UseUnixEpochBase.cs:36]: I think we should go with decimal with underscores: 621_355_968_000_000_000

**Review by csaba-sagi-sonarsource — APPROVED** [review]
Left a few nitpick comments. Once you address them I am fine with merging it.

File: analyzers/src/SonarAnalyzer.Common/Rules/UseUnixEpochBase.cs
Comment (cristian-ambrosini-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Rules/UseUnixEpochBase.cs:37]: This comment was about not forgetting to add once again that test case. I failed :D. Thanks for noticing it I'll fix it immediately

**Pull request merged**
